From d2f161471074d7c32540609ab6e6e9cda7083248 Mon Sep 17 00:00:00 2001
From: Flopster101 <nahuelgomez329@gmail.com>
Date: Sun, 5 Oct 2025 22:43:38 -0300
Subject: [PATCH] Launcher3: Support multiple wallpaper pickers

This new change makes it:

* Try com.android.wallpaper first,
* if not present, it will fall back to com.google.android.apps.wallpaper.
* if none are present, it will let the user choose.

Signed-off-by: Flopster101 <nahuelgomez329@gmail.com>
---
 .../launcher3/views/OptionsPopupView.java     | 31 ++++++++++++++++---
 1 file changed, 26 insertions(+), 5 deletions(-)

diff --git a/src/com/android/launcher3/views/OptionsPopupView.java b/src/com/android/launcher3/views/OptionsPopupView.java
index 5de488e..20011df 100644
--- a/src/com/android/launcher3/views/OptionsPopupView.java
+++ b/src/com/android/launcher3/views/OptionsPopupView.java
@@ -282,17 +282,38 @@ public class OptionsPopupView<T extends Context & ActivityContext> extends Arrow
             Toast.makeText(launcher, message, Toast.LENGTH_SHORT).show();
             return false;
         }
-        Intent intent = new Intent(Intent.ACTION_SET_WALLPAPER)
+
+        // Try multiple wallpaper picker packages in order of preference
+        String[] wallpaperPackages = {
+            launcher.getString(R.string.wallpaper_picker_package), // Custom package from config
+            "com.android.wallpaper", // AOSP wallpaper picker
+            "com.google.android.apps.wallpaper" // Google wallpaper picker
+        };
+
+        Intent baseIntent = new Intent(Intent.ACTION_SET_WALLPAPER)
                 .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
                 .putExtra(EXTRA_WALLPAPER_OFFSET,
                         launcher.getWorkspace().getWallpaperOffsetForCenterPage())
                 .putExtra(EXTRA_WALLPAPER_LAUNCH_SOURCE, "app_launched_launcher")
                 .putExtra(EXTRA_WALLPAPER_FLAVOR, "focus_wallpaper");
-        String pickerPackage = launcher.getString(R.string.wallpaper_picker_package);
-        if (!TextUtils.isEmpty(pickerPackage)) {
-            intent.setPackage(pickerPackage);
+
+        // Try each package until one works
+        for (String packageName : wallpaperPackages) {
+            if (TextUtils.isEmpty(packageName)) {
+                continue;
+            }
+
+            Intent intent = new Intent(baseIntent);
+            intent.setPackage(packageName);
+
+            // Check if the package can handle this intent
+            if (launcher.getPackageManager().resolveActivity(intent, 0) != null) {
+                return launcher.startActivitySafely(v, intent, placeholderInfo(intent)) != null;
+            }
         }
-        return launcher.startActivitySafely(v, intent, placeholderInfo(intent)) != null;
+
+        // If no specific package works, try without specifying a package (let system choose)
+        return launcher.startActivitySafely(v, baseIntent, placeholderInfo(baseIntent)) != null;
     }
 
     static WorkspaceItemInfo placeholderInfo(Intent intent) {
-- 
2.48.1

